nixos/wireless: enable PMF by default

rnhmjoj opened the pull request:
###### Motivation for this change

Alternative solution to PR #152443.
This fixes authentication failures to WPA3 networks (issue #151729)
by enabling protected management frames.

Note: old client without 802.11w support will still fail.

###### Things done

- [x] Tested connecting to SAE, Mixed WPA-PSK/SAE network
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

rnhmjoj committed:
nixos/wireless: enable PMF by default

Alternative solution to PR #152443.
This fixes authentication failures to WPA3 networks (issue #151729)
by enabling protected management frames.
Note: old client without 802.11w support will still fail.

rnhmjoj commented:
@mweinelt @blocktrron

mweinelt commented:
I'm confused.

Is this blanket breakage of older stations, or will they continue working fine with networks that don't require it? Setting `pmf=1` unconditionally in there is a difficult spot for a user to revert that change locally fwiw.

rnhmjoj commented:
> Is this blanket breakage of older stations, or will they continue working fine with networks that don't require it? Setting pmf=1 unconditionally in there is a difficult spot for a user to revert that change locally fwiw.

No, the only consequence of `pmf=1` is allowing connections to networks that require 802.11w.
Older clients will still fail on WPA3 networks, but work normally otherwise. At least, this is what I'm seeing from testing with a bunch of devices.

rnhmjoj committed:
nixos/wireless: implement opportunistic WPA3

It turns out it's actually possible to fall back to WPA2 in case the
authentication fails with WPA3. This was suggested to me in the hostapd
mailing list: add another network block with only WPA2 and lower
priority, for each network with WPA3. For clients with missing/broken
WPA3, wpa_supplicant will:

1. try the network block with higher priority first
2. fail and temporarily disable the network block
3. try the fallback network block and connect

This takes a little more time (still <5s) because wpa_supplicant
retries a couple times before disabling the network block, but it allows
old client to gracefully fall back to WPA2 on mixed WPA2/WPA3 networks.

To avoid downgrade attacks, clients with proper WPA3 should disable
this; in the future we may want to disable this option by default.

rnhmjoj commented:
I think I've found a solution to keep backward compatibility without sacrificing security too much.

rnhmjoj commented on nixos/modules/services/networking/wpa_supplicant.nix:
```diff
@@ -175,6 +202,18 @@ in {
         '';
       };
 
+      opportunisticWPA3 = mkOption {
```
Yeah, good idea.

rnhmjoj replied:
Name updated!

tfc commented:
I just checked this branch out on my Lenovo x250 and i can confirm now that this **fixes** my issue (https://github.com/NixOS/nixpkgs/issues/151729).

rnhmjoj commented:
I believe @tfc tested this with modern hardware with PMF. It would be good if someone other than me checked the fallback works by running this on an old network card.

rnhmjoj commented:
It's been a while but I haven't managed to get anyone else to test this change.
I'm merging it now anyway because I don't want to leave WPA3 (completely) broken any longer.
